tutorials/arvo/arvo.md

```diff
@@ -0,0 +1,519 @@
+modified copy of its context. The `move`s are used to interact with other vanes,
+while the new context allows the vane to save its state. The next time Arvo
+activates the vane it will have this context as its subject.
+
+This cycle of `%pass`ing a `note` to `%pass`ing a `task` to `%give`ing a `gift`
+to `%give`ing a `%sign` is summarized in the following diagram:
+
+<div style="text-align:center">
+<img class="mv5 w-150" src="https://media.urbit.org/docs/arvo/cycle.png">
+</div>
+
+Note that `%pass`ing a `note` doesn't _always_ result in a return - this diagram
+just
+shows the complete cycle. However, `%give`ing a `gift` is always in response to being
+`%pass`ed some `task`. Since the Arvo kernel acts as a middleman between all `move`s in Arvo, in diagrams we will
+generally represent the intermediate steps of a vane `%pass`ing a `note` to the kernel addressed to
+another vane followed by the kernel `%pass`ing a `task` to the
+addressed vane as a single arrow from one vane to the other to make the diagrams
+less cluttered.
+
+This overview has detailed how to pass a `card` to a particular vane. To see the
+`card`s each vane can be `%pass`ed as a `task` or return as a `gift` (as well as
+the semantics tied to them), each vane's public interface is explained in detail
+in its respective overview.
+
+
+#### `ovum`
+
+This mold is used to represent both steps and actions.
+
+A pair of a `wire` and a `curd`, with a `curd` being like a typeless `card`. The
+reason for a typeless `card` is that this is the data structure which Arvo uses
+to communicate with the runtime, and Unix events have no type. Additionally,
+upgrading the kernel may alter the type system and thus may not be able to be
+described within the current type system. Then the `wire` here is the default Unix `wire`, namely `//`. In particular, it is not a `duct` because `ovum`s come from the runtime rather than from within Arvo.
+
+### Arvo cores
+
+`arvo.hoon` has four additional cores that encode the functionality of Arvo. The [larval core](#larval-stage-core), [structural interface core](#structural-interface-core), and [implementation core](#implementation-core) each have five arms, called `+come`, `+load`, `+peek`, `+poke`, and `+wish`. Of these five arms, only `+poke` affects the [Arvo state](#the-state), while the rest leave the state invariant. Thus `+poke` is the aforementioned transition function that sends Arvo from one state to the next.
+
+A short summary of the purpose of each these arms are as follows:
+
+ - `+poke` is the transition function that `move`s Arvo from one state to the
+   next. It is the most fundamental arm in the entire system. It is a typed
+   transactional message that is delivered exactly once. This is sometimes said
+   to be impossible, and it is for standard operating systems, but that is not the case for single-level stores engaged in
+   a permanent session, as is the case among Arvo ships.
```

Comment:
The exactly-once guarantee is for Ames messages specifically, and it's built inside Ames, not Arvo.  Arvo only guarantees at-most-once processing of events (it could crash in the middle, in which case we throw it away).  Ames guarantees exactly-once through a system of acknowledgments on top of this (which only works because Arvo is a single-level store) and retrying.